state/machine_test.go

```diff
@@ -748,6 +748,63 @@ func (s *MachineSuite) TestMachineInstanceIdBlank(c *gc.C) {
 	c.Assert(string(iid), gc.Equals, "")
 }
 
+func (s *MachineSuite) TestAllSpacesNone(c *gc.C) {
+	machine, err := s.State.AddMachine("quantal", state.JobHostUnits)
+	spaces, err := machine.AllSpaces()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(spaces.SortedValues(), gc.DeepEquals, []string{})
+}
+
+func (s *MachineSuite) TestAllSpacesSimpleConstraints(c *gc.C) {
+	machine, err := s.State.AddMachine("quantal", state.JobHostUnits)
+	c.Assert(err, jc.ErrorIsNil)
+	err = machine.SetConstraints(constraints.Value{
+		Spaces: &[]string{"foo", "bar", "^baz"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	spaces, err := machine.AllSpaces()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(spaces.SortedValues(), gc.DeepEquals, []string{"bar", "foo"})
+}
+
+func (s *MachineSuite) TestAllSpacesEndpoints(c *gc.C) {
+	_, err := s.State.AddSpace("space", "", nil, true)
```

Comment:
A minor nit: it's nice to have the space names be concrete (e.g., dmz, db, internet), particularly as we later reference mysql. I think it also helps people understand spaces when seen in conjunction with `AddServiceWithBindings(<space-name>)`.